### PR TITLE
Direct IO

### PR DESCRIPTION
**Pull request by shikharish**
Related to #269.

### CODE REVIEW

**Comment by shikharish** [comment]
Still WIP.
@jesperpedersen Can you take a look, to see if I am heading in the right direction?

File: src/include/io.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#define PGMONETA_IO_H
+
+#ifdef __cpluscplus
+extern "C" {
+#endif
+
+#include <stdio.h>
+
+FILE*
+pgmoneta_open_file(const char* path, const char* modes);
```
Comment (jesperpedersen) [inline src/include/io.h:39]: Make `FILE*` an output parameter, and have `int` as the return value for success / failure

File: src/include/io.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+int
+pgmoneta_write_file(void* buffer, size_t size, size_t n, FILE* file);
```
Comment (jesperpedersen) [inline src/include/io.h:42]: Put `FILE*` as first parameter.

Do we need `n` ?

File: src/include/io.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
```
Comment (jesperpedersen) [inline src/include/io.h:43]: We need read and close as well

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/include/io.h (hunk 2 above)
Comment (shikharish) [reply src/include/io.h:42]: How would we know the total number of bytes without `n`?

Comment (jesperpedersen) [reply src/include/io.h:42]: Just use `size` - then partition internally in the function.

... provided that it will work ... otherwise just keep it

**Comment by jesperpedersen** [comment]
@shikharish Are you doing ok on this ?

**Comment by shikharish** [comment]
Yes, working on it. Will share an update soon.

**Comment by jesperpedersen** [comment]
@shikharish Are you doing good on this ?

File: src/include/io.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#define PGMONETA_IO_H
+
+#include <sys/types.h>
+#ifdef __cpluscplus
+extern "C" {
+#endif
+
+#include <fcntl.h>
+#include <stdio.h>
+#define BLOCK_SIZE 512
```
Comment (jesperpedersen) [inline src/include/io.h:41]: Separate with a line feed

File: src/include/io.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+int pgmoneta_open_file(const char* path, int oflag);
```
Comment (jesperpedersen) [inline src/include/io.h:43]: Missing documentation and formatting. Remove `const` as well

File: src/include/io.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+ssize_t pgmoneta_write_file(void* buffer, size_t bytes, int fd);
```
Comment (jesperpedersen) [inline src/include/io.h:45]: Same

File: src/include/io.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+void*pgmoneta_alligned_malloc(size_t size);
```
Comment (jesperpedersen) [inline src/include/io.h:47]: Same

File: src/include/io.h (hunk 6 above)
Comment (jesperpedersen) [reply src/include/io.h:45]: Move `fd` to the first parameter

File: src/include/io.h (hunk 5 above)
Comment (jesperpedersen) [reply src/include/io.h:43]: oflag -> flags

File: src/libpgmoneta/archive.c [hunk 8]
```diff
@@ -27,6 +27,7 @@
  */
 
 /* pgmoneta */
+#include <io.h>
```
Comment (jesperpedersen) [inline src/libpgmoneta/archive.c:30]: This need to be sorted below `<pgmoneta.h>` in that section

File: src/libpgmoneta/io.c [hunk 9]
```diff
@@ -0,0 +1,75 @@
+#include <io.h>
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:1]: Missing header

File: src/libpgmoneta/io.c [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+#include <logging.h>
+#include <stdint.h>
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:3]: Separate with a line feed

File: src/libpgmoneta/io.c [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+int
+pgmoneta_open_file(const char* path, int oflag)
+{
+   int fd = open(path, oflag, 0600);
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:11]: Can be on one line

File: src/libpgmoneta/io.c [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+{
+   ssize_t bytes_written = 0;
+   int flags = fcntl(fd, F_GETFL);
+   if ((uintptr_t)buffer % BLOCK_SIZE != 0 && flags & O_DIRECT)
+   {
+      void* alligned_buf =
+         pgmoneta_alligned_malloc(((bytes) / BLOCK_SIZE + 1) * BLOCK_SIZE);
+      memcpy(alligned_buf, buffer, bytes);
+
+      // pgmoneta_log_debug("NOT ALLIGNED & O_DIRECT is enabled");
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:27]: Comment

File: src/libpgmoneta/io.c [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+         }
+      }
+
+      bytes_written = write(fd, alligned_buf, bytes);
+
+      if (bytes % BLOCK_SIZE != 0)
+      {
+         if (fcntl(fd, F_SETFL, flags))
+         {
+            // pgmoneta_log_debug("failed to re-enable O_DIRECT");
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:43]: Missing something ?

File: src/libpgmoneta/io.c [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+   }
+   else
+   {
+      bytes_written = write(fd, buffer, bytes);
+   }
+
+   if (bytes_written < 0)
+   {
+      // pgmoneta_log_debug("error writing to fd: %d", fd);
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:55]: You can always use `pgmoneta_log_trace()` for stuff like this

File: src/libpgmoneta/io.c [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+      return -1;
+   }
+
+   // pgmoneta_log_debug("write: +%ld bytes fd: %d", bytes_written, fd);
+   return bytes_written;
+}
+
+// ssize_t pgmoneta_read_file(void *buffer, size_t buf_len, int fd) {
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:63]: Do we need this one ?

File: src/libpgmoneta/memory.c [hunk 16]
```diff
@@ -27,6 +27,7 @@
  */
 
 /* pgmoneta */
+#include "io.h"
```
Comment (jesperpedersen) [inline src/libpgmoneta/memory.c:30]: Sorting again

File: src/libpgmoneta/message.c [hunk 17]
```diff
@@ -27,6 +27,7 @@
  */
 
 /* pgmoneta */
+#include <io.h>
```
Comment (jesperpedersen) [inline src/libpgmoneta/message.c:30]: Sorting again

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
General comment: alligned -> aligned

**Comment by jesperpedersen** [comment]
You likely need `open(..., ... | O_SYNC);` as well

**Comment by shikharish** [comment]
@jesperpedersen I think we only need the write function. Currently pgmoneta_open_file() function is just a wrapper around open() and does not serve any significant purpose. I think it can be removed...

**Comment by jesperpedersen** [comment]
@shikharish CI for MacOS

File: src/libpgmoneta/archive.c [hunk 18]
```diff
@@ -32,11 +32,13 @@
 #include <deque.h>
 #include <gzip_compression.h>
 #include <info.h>
+#include <io.h>
 #include <logging.h>
 #include <lz4_compression.h>
 #include <management.h>
 #include <network.h>
 #include <restore.h>
+#include <unistd.h>
```
Comment (jesperpedersen) [inline src/libpgmoneta/archive.c:41]: This should go under the system includes

File: src/libpgmoneta/io.c [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+      if (odd_bytes != 0)
+      {
+         char* tmp_buf = (char*)buffer;
+         while (bytes_written + 8192 < bytes)
+         {
+            bytes_written += write(fd, tmp_buf + bytes_written, 8192);
+         }
+         if (fcntl(fd, F_SETFL, flags & ~O_DIRECT))
+         {
+            pgmoneta_log_error("failed to disable O_DIRECT");
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:67]: failed -> Failed

File: src/libpgmoneta/io.c [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+            return -1;
+         }
+
+         bytes_written += write(fd, tmp_buf + bytes_written, bytes - bytes_written);
+
+         if (fcntl(fd, F_SETFL, flags))
+         {
+            pgmoneta_log_error("failed to re-enable O_DIRECT");
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:75]: failed -> Failed

File: src/libpgmoneta/io.c [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+      bytes_written = write(fd, buffer, bytes);
+   }
+
+   if (bytes_written < 0)
+   {
+      pgmoneta_log_debug("error writing to fd: %d", fd);
+      return -1;
+   }
+
+   pgmoneta_log_debug("write: +%ld bytes fd: %d", bytes_written, fd);
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:96]: debug -> trace

File: src/libpgmoneta/io.c [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+   }
+   else
+   {
+      bytes_written = write(fd, buffer, bytes);
+   }
+
+   if (bytes_written < 0)
+   {
+      pgmoneta_log_debug("error writing to fd: %d", fd);
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:92]: Should be an error, and error -> Error.

You should include the the error string, and then clear errno

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by shikharish** [comment]
> @shikharish CI for MacOS

O_DIRECT is only present on Linux. MacOS has something similar, F_NOCACHE. Should there be a separate implementation for Macos or we just use buffered-io?

**Comment by jesperpedersen** [comment]
Nah, lets just use `#ifdef HAVE_LINUX` and a general `#else` with buffered I/O.

Did you create a test application that

1) Delete the test file
2) Write (O_DIRECT | O_SYNC) the big test file w/ timing
3) Read (O_DIRECT | O_SYNC) the big test file w/ timing
4) Delete the test file
5) Buffered write the big test file w/ timing
6) Buffered write the big test file w/ timing
7) Delete the test file

Make sure that test file is *much* bigger than the kernel disk cache... and you should probably do an average over multiple runs.

You can compare the results you are getting with tools like `dd` - plus the maximum speed of your hard drive

**Comment by jesperpedersen** [comment]
Rebase

**Comment by shikharish** [comment]
@jesperpedersen I have written a test file in [this](https://github.com/shikharish/direct-io) repo. It uses the same implementation of io.c. I am getting slightly worse write speeds and better read speeds for direct-io compared to buffered-io. 

Here is the statistics for writing a 12GB file with a 256KB buffer:
```
****DIRECT-IO TEST****
file size = 12GB, buffer size = 256KB
----RUN 1----
bytes written= 12884901888      write time= 11.477559 sec
bytes read= 12884901888 read time = 6.245287 sec
----RUN 2----
bytes written= 12884901888      write time= 11.939732 sec
bytes read= 12884901888 read time = 5.429730 sec
----RUN 3----
bytes written= 12884901888      write time= 11.389632 sec
bytes read= 12884901888 read time = 5.826528 sec
----RUN 4----
bytes written= 12884901888      write time= 12.695171 sec
bytes read= 12884901888 read time = 6.625986 sec
----RUN 5----
bytes written= 12884901888      write time= 11.689396 sec
bytes read= 12884901888 read time = 6.390608 sec
--------------
AVG WRITE TIME = 11.838298 sec   AVG WRITE SPEED = 1037.987061 MB/s
AVG READ TIME = 6.103628 sec     AVG READ SPEED = 2013.229004 MB/s
**********************
./test  0.26s user 8.03s system 9% cpu 1:29.71 total


****BUFFERED-IO TEST****
file size = 12GB, buffer size = 256KB
----RUN 1----
bytes written= 12884901888      write time= 11.035978 sec
bytes read= 12884901888 read time = 7.788012 sec
----RUN 2----
bytes written= 12884901888      write time= 10.380377 sec
bytes read= 12884901888 read time = 6.949536 sec
----RUN 3----
bytes written= 12884901888      write time= 9.475489 sec
bytes read= 12884901888 read time = 7.588753 sec
----RUN 4----
bytes written= 12884901888      write time= 11.105805 sec
bytes read= 12884901888 read time = 7.063450 sec
----RUN 5----
bytes written= 12884901888      write time= 9.865578 sec
bytes read= 12884901888 read time = 7.256378 sec
--------------
AVG WRITE TIME = 10.372645 sec   AVG WRITE SPEED = 1184.654419 MB/s
AVG READ TIME = 7.329226 sec     AVG READ SPEED = 1676.575439 MB/s
**********************
./test  0.22s user 43.74s system 49% cpu 1:29.54 total
```

dd gives much slower speeds, the dd test file is also present in the repo.

**Comment by jesperpedersen** [comment]
That looks ok.

Take a look at the CI

File: src/include/io.h [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+#define BLOCK_SIZE 512
+
+/**
+ * Wrapper around write() libc call
+ * @param fd The file descriptor
+ * @param buffer The buffer to be written
+ * @param bytes The number of bytes to be written
+ * @return The number of bytes written, -1 on error
+ */
+ssize_t pgmoneta_write_file(int fd, void* buffer, size_t bytes);
```
Comment (jesperpedersen) [inline src/include/io.h:51]: Formatting

File: src/include/io.h [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Wrapper around aligned_alloc() libc call
+ * @param size The size of the buffer
+ */
+void*pgmoneta_aligned_malloc(size_t size);
```
Comment (jesperpedersen) [inline src/include/io.h:57]: Formatting

File: src/main.c [hunk 25]
```diff
@@ -986,7 +986,7 @@ accept_mgt_cb(struct ev_loop* loop, struct ev_io* watcher, int revents)
             result = pgmoneta_delete(srv, backup_id);
             pgmoneta_log_trace("Delete: %d", result);
             pgmoneta_delete_wal(srv);
-            pgmoneta_management_write_delete(NULL, client_fd, srv /*, result */);
+            pgmoneta_management_write_delete(NULL, client_fd, srv /*, result */ );
```
Comment (jesperpedersen) [inline src/main.c:989]: We can clean this up

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
You need to verify valgrind too - see `null_buffer` f.ex

**Comment by jesperpedersen** [comment]
Please, check CI and your local runs.

Maybe use clang and Debug builds to get closer

```
export CC=/usr/bin/clang && cmake -DCMAKE_BUILD_TYPE=Debug ..
```

**Comment by shikharish** [comment]
@jesperpedersen Don't see any errors on my local build. Any idea how I can reproduce this?

**Comment by jesperpedersen** [comment]
You have the stack traces from the CI to work from -- otherwise https://github.com/pgmoneta/pgmoneta/tree/main/contrib/valgrind

File: src/libpgmoneta/io.c [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+         void* aligned_buf =
+            pgmoneta_aligned_malloc((bytes / BLOCK_SIZE + 1) * BLOCK_SIZE);
+         memcpy(aligned_buf, buffer, bytes);
+         buffer = aligned_buf;
+      }
+
+      if (odd_bytes != 0)
+      {
+         char* tmp_buf = (char*)buffer;
+         while (bytes_written + 8192 < bytes)
```
Comment (jesperpedersen) [inline src/libpgmoneta/io.c:61]: Have you looked at https://eklitzke.org/efficient-file-copying-on-linux ?

Maybe for now we should just start with `#define WRITE_SIZE 8192`...

Comment (jesperpedersen) [reply src/libpgmoneta/io.c]: Maybe try and see how different sizes affects your test setup

Comment (jesperpedersen) [reply src/libpgmoneta/io.c]: Maybe an idea is to read the system buffer size into a `disk_buffer_size` and use that - default would be 8192

File: src/include/io.h [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+extern "C" {
+#endif
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/types.h>
+
+#define BLOCK_SIZE 512
+
+#ifndef HAVE_LINUX
```
Comment (jesperpedersen) [inline src/include/io.h:44]: FreeBSD has O_DIRECT as well

**Comment by jesperpedersen** [comment]
What happened to `O_DIRECT | O_SYNC` ?

File: src/include/io.h (hunk 27 above)
Comment (jesperpedersen) [reply src/include/io.h:44]: Use virt-manager to setup a FreeBSD 14.1 VM

File: src/libpgmoneta/io.c (hunk 26 above)
Comment (shikharish) [reply src/libpgmoneta/io.c]: 128KB size buffer works best on my setup too. I have hardcoded this for now. I couldn't find a way to find the disk buffer size using libc calls. Maybe I can make a shell call to `blockdev` and find it.

**Comment by jesperpedersen** [comment]
Merged.

Thanks for your contribution !

**Pull request merged**